lib/Parse/ParseExpr.cpp

```diff
@@ -1913,7 +1913,11 @@ ParserResult<Expr> Parser::parseExprStringLiteral() {
 
   // The start location of the entire string literal.
   SourceLoc Loc = Tok.getLoc();
-  SourceLoc EndLoc = Loc.getAdvancedLoc(Tok.getLength());
+  // Put the close brace of the of the body of the AppendingExpr just inside
+  // of the literal so that it doesn't go past where a missing close brace
+  // for an enclosing IterableTypeDecl will be added.
+  // (See \c parseMatchingToken.)
+  SourceLoc EndLoc = Loc.getAdvancedLoc(Tok.getLength() - 1);
```

Comment:
This `EndLoc` that is used for the "closing brace" is character-based, it should actually be token based (it was character-based before as well, before this change, but we should correct it).

It's not clear to me if the implicit `BraceStmt` that is created for the `TapExpr` is supposed to be the `SourceRange` of the whole string literal or just the range of the expression segments, but if it is supposed to be the SourceRange of the whole string literal then it should be matching the string literal's SourceRange, whose start and end locations are the token location of where the literal string begins.

Reply:
Sounds good; I'll try that. As you saw, before this change it was one character past the end of the token; neither fish nor fowl.

Reply:
I think it will work better to point at the last token in the AppendingExpression. Going to try that.